packages/workspace/src/command-line/project-graph/build-dependencies/explicit-project-dependencies.ts

```diff
@@ -29,15 +35,32 @@ export function buildExplicitTypeScriptDependencies(
     });
   });
 
-  function findTargetProject(importExpr, nodes) {
-    return Object.keys(nodes).find(projectName => {
-      const p = nodes[projectName];
-      const normalizedRoot = normalizedProjectRoot(p);
-      return (
-        importExpr === `@${ctx.nxJson.npmScope}/${normalizedRoot}` ||
-        importExpr.startsWith(`@${ctx.nxJson.npmScope}/${normalizedRoot}#`) ||
-        importExpr.startsWith(`@${ctx.nxJson.npmScope}/${normalizedRoot}/`)
-      );
-    });
+  function findTargetProject(importExpr, nodes, filePath) {
+    compilerHost = compilerHost || getCompilerHost();
+    const { options, host } = compilerHost;
+
+    const { resolvedModule } = ts.resolveModuleName(
+      importExpr,
+      filePath,
+      options,
+      host
+    );
+
+    // module was not found in the workspace - return
+    if (!resolvedModule) {
+      return;
+    }
+
+    return Object.keys(nodes).find(projectName =>
+      resolvedModule.resolvedFileName.includes(
+        nodes[projectName].data.sourceRoot
```

Comment:
Use `root` instead. The file doesn't have to be in the `sourceRoot`.

Reply:
I wanted to use `sourceRoot` because it's more unique than just `root`. For example two libraries, one nested in each other (`shared/product`, `shared/product/data`) with the following structure:
```tree
libs
└── shared
    └── product
        ├── README.md
        ├── data
        │   ├── README.md
        │   ├── jest.config.js
        │   ├── src
        │   │   ├── index.ts
        │   │   ├── lib
        │   │   │   ├── shared-product-data.module.spec.ts
        │   │   │   └── shared-product-data.module.ts
        │   │   └── test-setup.ts
        │   ├── tsconfig.json
        │   ├── tsconfig.lib.json
        │   ├── tsconfig.spec.json
        │   └── tslint.json
        ├── jest.config.js
        ├── src
        │   ├── index.ts
        │   ├── lib
        │   │   ├── shared-product.module.spec.ts
        │   │   └── shared-product.module.ts
        │   └── test-setup.ts
        ├── tsconfig.json
        ├── tsconfig.lib.json
        ├── tsconfig.spec.json
        └── tslint.json
```

If we have an import that resolves to `shared/product/data/src/index.ts`, then the root of `shared/product` could be true as well because shared/product appears in shared/product/data. 

Using the `sourceRoot` would make it unique enough to know it only belongs to that library. 

Hopefully devs would follow the structure pointed out in the `workspace.json` to actually put their source in the `src` directory.

I'm ok with having the check with the root, but then we would need to have a guarantee that the most deeply nested library always appear first when generating the projects graph nodes.

Reply:
Well, it looks like the nodes are sorted by length 😄 
https://github.com/nrwl/nx/blob/a82737968b698ac43da04e66614fc0877869fe9c/packages/workspace/src/command-line/project-graph/build-nodes/workspace-projects.ts#L37-L42

I'll change it.